# Q1 a, b

In [2]:
# Q1 a,b 
import pandas as pd

data_Q3 = pd.ExcelFile('./PS5_Q3_data.xls')
sheet_name = 'Returns of various assets'
data = pd.read_excel(data_Q3, sheet_name=sheet_name, index_col=0, header=1).dropna(axis=1)
FS_ret = data.iloc[:, 0]
SP_ret = data.iloc[:, 1]

E_rFS = FS_ret.mean()
E_rSP = SP_ret.mean()

sigma_FS = FS_ret.std()
sigma_SP = SP_ret.std()

rho = FS_ret.corr(SP_ret)

r_f = 0.0033

beta_FS = (rho * sigma_FS) / sigma_SP

E_rFS_CAPM = r_f + beta_FS * (E_rSP - r_f)

alpha = E_rFS - E_rFS_CAPM

print(f"Beta: {beta_FS}")
print(f"Expected Return: {E_rFS_CAPM}")
print(f"alpha: {alpha}")

Beta: 0.05576456806668122
Expected Return: 0.00358341189823299
alpha: 0.005975381205215284


### Beta: 0.05576
### Expected Return: 0.003583
### alpha: 0.005975


# Q2 a, b, c, d

In [5]:
# Q2 
import numpy as np
E_ri = 0.15  
sigma_i = 0.60  
rho_ij = 0.5  
r_f = 0.10  
N = 25  
E_rp = E_ri  
sigma_p = np.sqrt((1 / N) * sigma_i**2 + (1 - 1 / N) * rho_ij * sigma_i**2)

target_sigma = 0.43 

def solve_for_N(target_sigma, sigma_i, rho_ij):
    N_required = (sigma_i**2 * (1 - rho_ij)) / (target_sigma**2 - rho_ij * sigma_i**2)
    return int(np.ceil(N_required))  

N_min = solve_for_N(target_sigma, sigma_i, rho_ij)

systematic_risk = np.sqrt(rho_ij) * sigma_i

CML_slope = (E_rp - r_f) / sigma_p

E_rp, sigma_p, N_min, systematic_risk, CML_slope

(0.15, 0.4326661530556787, 37, 0.4242640687119285, 0.11556254088025605)

### a) Expected return is 0.15, standard deviation is 0.432666
### b) The smallest number of stocks is 37
### c) Std is 0.42426
### d) Slope of CML is 0.11556

# Q3 a

In [9]:
# Q3 a)
beta_AAPL = 1.2314
beta_AMZN = 1.1331
beta_JNJ = 0.4898
beta_large_port = (beta_AAPL + beta_AMZN + beta_JNJ) / 3

market_risk_premium = 0.00551

rp_large_port = beta_large_port * market_risk_premium

rp_large_port

0.005242397666666667

In [20]:
# Q3 b)
var_aapl=0.052878611**2
var_amzn=0.072138722**2
var_jnj=0.04171887**2

var_portfolio_residual = (var_aapl + var_amzn + var_jnj) / 9
var_portfolio_residual

0.0010822896474621672

In [21]:
# Q3 c)
var_mkt =0.055563898
systematic_risk = beta_large_port**2 * var_mkt

non_systematic_risk = var_portfolio_residual

# Display the results
systematic_risk, non_systematic_risk

(0.050297851113494906, 0.0010822896474621672)

In [22]:
# Q3 d)
# MATX、PCRX、MMI
beta_MATX = 1.1091
beta_PCRX = 0.7679
beta_MMI= 1.2242
beta_small_port = (beta_MATX+ beta_PCRX + beta_MMI) / 3

market_risk_premium = 0.00551
rp_small_port = beta_small_port * market_risk_premium

rp_small_port


0.005695870666666667

In [23]:
# Q3 e)
market_risk_premium = 0.00551
expected_return_difference = (beta_small_port - beta_large_port) * market_risk_premium

expected_return_difference

0.0004534729999999996

In [26]:
#Q3 f)

var_matx_residual = 0.084623288 **2 
var_pcrx_residual = 0.187477171 **2 
var_mmi_residual =0.073949918 **2

var_small_portfolio_residual = (var_matx_residual + var_pcrx_residual + var_mmi_residual) / 9

var_small_portfolio = (beta_small_port**2 * var_mkt) + var_small_portfolio_residual
var_large_portfolio = (beta_large_port**2 * var_mkt) + var_portfolio_residual
var_zero_investment = var_small_portfolio + var_large_portfolio

var_zero_investment

0.11606457588133864

### a) risk premium: 0.005242
### b) residual variance: 0.001082
### c) Systematic risk is 0.0502979, non-systematic risk is 0.001082
### d) The monthly risk premium of this portfolio is 0.0056959
### e) E(rd- ra) = 0.000453473
### f) The variance of the zero-investment return is 0.1160

# Q4

In [8]:
#Q4
beta_m_matx = 1
beta_smb_matx = 0.4336
beta_hml_matx = 0.6614

beta_m_lanc = 0.2799
beta_smb_lanc = -0.0397
beta_hml_lanc = -0.1477

market_risk_premium_ff = 0.0065 
smb_premium = 0.00107 
hml_premium = 0.00301

beta_m_portfolio = (beta_m_matx + beta_m_lanc) / 2
beta_smb_portfolio = (beta_smb_matx + beta_smb_lanc) / 2
beta_hml_portfolio = (beta_hml_matx + beta_hml_lanc) / 2

monthly_risk_premium = (beta_m_portfolio * market_risk_premium_ff) + \
                       (beta_smb_portfolio * smb_premium) + \
                       (beta_hml_portfolio * hml_premium)

monthly_risk_premium

0.00514353

# Q5

In [5]:
import numpy as np

# Given data for expected returns and standard deviations
expected_returns = np.array([0.0550, 0.0491, 0.0437, 0.0658, 0.0779, 0.0211, 0.1024])
weights = np.array([0.05, 0.20, 0.10, 0.40, 0.05, 0.10, 0.10])

# Covariance matrix from the second image
cov_matrix = np.array([
    [0.027553, 0.017111, 0.018188, 0.026707, 0.019405, 0.016228, 0.019440],
    [0.017111, 0.016763, 0.015552, 0.017930, 0.012844, 0.011785, 0.016791],
    [0.018188, 0.015552, 0.020944, 0.018732, 0.014600, 0.012503, 0.016551],
    [0.026707, 0.017930, 0.018732, 0.030481, 0.021456, 0.016061, 0.021317],
    [0.019405, 0.012844, 0.014600, 0.021456, 0.026869, 0.010873, 0.018163],
    [0.016228, 0.011785, 0.012503, 0.016061, 0.010873, 0.014980, 0.012062],
    [0.019440, 0.016791, 0.016551, 0.021317, 0.018163, 0.012062, 0.023466]
])

expected_return_world = np.dot(weights, expected_returns)
variance_world = np.dot(weights.T, np.dot(cov_matrix, weights))
cov_japan_world = np.dot(cov_matrix[4], weights)

beta_japan = cov_japan_world / variance_world
expected_return_world, variance_world, cov_japan_world, beta_japan

(0.05950500000000001, 0.01982004, 0.0178285, 0.8995188707994535)

In [19]:
risk_free_rate = 0.005
expected_return_japan_capm = risk_free_rate + beta_japan * (expected_return_world - risk_free_rate)
expected_return_japan_capm

0.05402827605292422

### E(r_world) = 0.0595
### Cov(World, Japan) = 0.0178
### Var(World) = 0.0198
### Beta(Japan) = 0.8995
### E(r_Japan) (CAPM) =  0.05453

The expected return of JP stock index by Global CAPM is 5.403%, which is close to the real return 5.950%. And JP index's beta is close to 1, so they are corresponding.
